src/Microsoft.ML.Transforms/NAIndicatorTransform.cs

```diff
@@ -68,116 +78,196 @@ private static VersionInfo GetVersionInfo()
         {
             // Item type must have an NA value. We'll get the predicate again later when we're ready to use it.
             Delegate del;
             if (Conversions.Instance.TryGetIsNAPredicate(type.ItemType, out del))
                 return null;
             return string.Format("Type '{0}' is not supported by {1} since it doesn't have an NA value",
-                type, LoaderSignature);
+                type, LoadName);
+        }
+
+        public class ColumnInfo
+        {
+            public readonly string Input;
+            public readonly string Output;
+
+            /// <summary>
+            /// Describes how the transformer handles one column pair.
+            /// </summary>
+            /// <param name="input">Name of input column.</param>
+            /// <param name="output">Name of output column.</param>
+            public ColumnInfo(string input, string output)
+            {
+                Input = input;
+                Output = output;
+            }
         }
 
-        private const string RegistrationName = "NaIndicator";
+        private const string RegistrationName = nameof(NAIndicatorTransform);
+
+        // The input column types
+        private ColumnType[] _inputTypes;
+        // The output column types
+        private ColumnType[] _outputTypes;
 
-        // The output column types, parallel to Infos.
-        private readonly ColumnType[] _types;
+        private static (string input, string output)[] GetColumnPairs(ColumnInfo[] columns)
+        {
+            Contracts.CheckValue(columns, nameof(columns));
+            return columns.Select(x => (x.Input, x.Output)).ToArray();
+        }
+
+        protected override void CheckInputColumn(ISchema inputSchema, int col, int srcCol)
+        {
+            var type = inputSchema.GetColumnType(srcCol);
+            string reason = TestType(type);
+            if (reason != null)
+                throw Host.ExceptParam(nameof(inputSchema), reason);
+        }
+
+        private (ColumnType[], ColumnType[]) GetTypes(ISchema schema)
+        {
+            var inputTypes = new ColumnType[ColumnPairs.Length];
+            var outputTypes = new ColumnType[ColumnPairs.Length];
+            for (int i = 0; i < ColumnPairs.Length; i++)
+            {
+                schema.TryGetColumnIndex(ColumnPairs[i].input, out int colSrc);
+                var type = schema.GetColumnType(colSrc);
+
+                if (!type.IsVector)
+                {
+                    inputTypes[i] = type.AsPrimitive;
+                    outputTypes[i] = BoolType.Instance;
+                }
+                else
+                {
+                    inputTypes[i] = new VectorType(type.ItemType.AsPrimitive, type.AsVector);
+                    outputTypes[i] = new VectorType(BoolType.Instance, type.AsVector);
+                }
+            }
+            return (inputTypes, outputTypes);
+        }
 
         /// <summary>
```

Comment:
>summary [](start = 13, length = 7)

a) this summary should change, because it is not for the public facing API. 
Also the `param` comments are bad.

b) this constructor should not be public at all. The only way to train a trainable transformer should be via estimator's `Fit` call, not directly via some constructor

Reply:
Actually, I take it even further. This transformer should not be trainable at all. It should not memorize the column types, only names.
All of the column type stuff should belong to `Mapper`. 

---
In reply to: [224863446](https://github.com/dotnet/machinelearning/pull/1217#discussion_r224863446) [](ancestors = 224863446)